### PR TITLE
Restructure pipelines

### PR DESCRIPTION
**Pull request by JensWehner**
## Summary

Reduce unnecessary GitHub Actions runs by limiting workflows to `main` and
triggering them only when their relevant inputs change.

Split the existing examples pipeline by consumer so GauXC changes do not run
the direct LibTorch and FTorch integrations, and model-example changes do not
start the full GauXC matrix.

## Changes

- Restrict Tests, Model benchmarks, GauXC examples, and Model examples to:
  - pull requests targeting `main`
  - pushes to `main`
- Add workflow-specific path filters.
- Reuse each path list between `push` and `pull_request` with YAML anchors.
- Had to fix the cpp model example as it was not CI tested

### CODE REVIEW

**Review by tvogels — APPROVED** [review]

File: examples/cpp/cpp_integration/main.cpp [hunk 1]
```diff
@@ -95,12 +95,16 @@ get_exc_and_grad(const torch::jit::Method &exc_func, const FeatureDict &features
...
+  for (const auto& entry : features) {
+    std::string key = entry.key();
+    bool requires_grad = (key == "density" || key == "grad" || key == "kin" || key == "coarse_0_atomic_coords" || key == "grid_coords" || key == "grid_weights");
+
+    auto tensor = entry.value().clone().requires_grad_(requires_grad);
+    if (requires_grad) {
+      input_tensors.push_back(tensor);
+      tensor_keys.push_back(key);
+    }
+    features_with_grad.insert(key, tensor);
```
Comment (tvogels) [inline examples/cpp/cpp_integration/main.cpp:107]: What's the change here?

Comment (JensWehner) [reply examples/cpp/cpp_integration/main.cpp:107]: before all features required a grad, which crashed the model. 

Some of the features do not need a grad, that has been adjusted.

Comment (awvwgk) [reply examples/cpp/cpp_integration/main.cpp:107]: This avoids to compute gradients for the `atomic_*` features as far as I can see.

**Review by awvwgk — APPROVED** [review]
Thanks for cleaning those up.
